controllers/ideaController.js

```diff
@@ -12,77 +20,185 @@ exports.createIdea = function(req,res){
+			res.send(util.wrapBody(result));
+		}
+	});
 
-			var options = {
-				pageNum:pageNum,
-				pageSize:pageSize
-			}
+};
 
-			ideaRepostory.queryIdea(options,function(err,result){
-				if (err) {
-					console.log(err);
-					res.send(util.wrapBody('Internal Err','E'));
-				} else {
-					res.send(util.wrapBody(result));
+exports.publishIdea = function(req,res){
+	var userId = req.token.userId;
+	var ideaId = req.params.id;
+
+	var data = req.body;
+	var idea = null;
+	var newIdea = null;
+	var errMsg = null;
+
+	const STATE_CREATE_IDEA = 1;
+	const STATE_UPDATE_IDEA = 2;
+ 	const STATE_CHECK_REQUIRED_DATA = 3;
+	const STATE_PUBLISH_IDEA = 4;
+	const STATE_SEND_RESPONSE = 0;
+
+	if (ideaId == 'new') {
+		stateMachine(null,STATE_CREATE_IDEA);
+	}else{
+		stateMachine(null,STATE_UPDATE_IDEA);
+	}
+
+	function stateMachine(err,toState){
+		console.log('state:',toState);
+
+		if (err) {
+			console.log('error:',err);
+			errMsg = 'Internal Error';
+			toState = STATE_SEND_RESPONSE;
+		}
+
+		switch(toState){
+			case STATE_CREATE_IDEA:
+				data.innovator = userId;
+				ideaRepostory.create(data,function(err,result){
+					idea = result;
+					stateMachine(err,STATE_CHECK_REQUIRED_DATA);
+				});
+			break;
+			case STATE_UPDATE_IDEA:
+				ideaRepostory.update({
+					_id:ideaId,
+					innovator:userId
+				},data,function(err,result){
+					idea = result;
+					stateMachine(err,STATE_CHECK_REQUIRED_DATA);
+				});
+			break;
+			case STATE_CHECK_REQUIRED_DATA:
+
+				if (!idea) {
+					errMsg = 'Invalid innovator';
+					stateMachine(null,STATE_SEND_RESPONSE);
+				}else{
+					console.log(idea);
+
+					if(checkIdeaComplete(idea)){
+						stateMachine(null,STATE_PUBLISH_IDEA);
+					}else{
+						errMsg = 'Miss data to publish';
+						stateMachine(null,STATE_SEND_RESPONSE);
+					}
 				}
-			});
+
+				
+			break;
+			case STATE_PUBLISH_IDEA:
+				ideaRepostory.update({
+					_id:idea._id
+				},{
+					state:'published'
+				},function(err,result){
+					newIdea = result;
+					stateMachine(err,STATE_SEND_RESPONSE);
+				});
+			break;
+			case STATE_SEND_RESPONSE:
+				if (errMsg) {
+					res.send(util.wrapBody(errMsg,'E'));
+				}else{
+					res.send(util.wrapBody({idea:newIdea}));
+				}
+			break;
+			default:
+				console.log('Invalid State');
+				errMsg = 'Internal Error';
+				stateMachine(null,STATE_SEND_RESPONSE);
 		}
-	})
+		
+	}
+};
+
+function checkIdeaComplete(idea){
+	var requires = ['name','background','solution','innovator',
+					'sector','painPoint','hrRequirement'];
+
+	for (var i = requires.length - 1; i >= 0; i--) {
+		if(!idea[requires[i]]){
+			console.log('Miss ' + requires[i]);
+			return false;
+		}
+	}
 
+	return true;
 }
 
-
 exports.updateIdea = function(req,res){
-	var idea = req.body.idea;
 
-	ideaRepostory.update(idea,function(err,result){
+	var userId = req.token.userId;
+	var ideaId = req.params.id;
+
+	var idea = req.body;
+
+	if ('_id' in idea) delete idea._id;
+	if ('state' in idea) delete idea.state;
```

Comment:
I do not understand why delete state here. How are you about to manage the state? It's business logic, should be processed at least at controller's level. do not bury it to repository.